File: stack/generateParenthesis.py
class Solution:
    def __init__(self):
        self.res = []

    def addParentheses(self, o: int, c: int, s: str):
        if not o and not c:
            self.res.append(s)
            return 
        elif o == c:
            self.addParentheses(o - 1, c, s + "(")
        elif not o and c:
            self.addParentheses(o, c - 1, s + ")")
        else:
            self.addParentheses(o - 1, c, s + "(")
            self.addParentheses(o, c - 1, s + ")")

    def generateParenthesis(self, n: int) -> List[str]:
        self.addParentheses(n, n, "")
        return self.res  

File: stack/generateParenthesis.cpp
class Solution {
public:
    void bt(int o, int c, string stack, vector<string>& res){
        if (o == 0 && c == 0){
            res.push_back(stack);
            return;
        }

        if(o == c)
            bt(o - 1, c, stack + '(', res);
        else if(o == 0)
            bt(o, c - 1, stack + ')', res);
        else if(o < c){
            bt(o - 1, c, stack + '(', res);
            bt(o, c - 1, stack + ')', res);
        }
    }
    
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        string stack;
        bt(n, n, stack, res);
        return res;
    }
};
